salt/modules/x509.py

```diff
@@ -1392,18 +1393,43 @@ def create_certificate(
                 ['listen_in', 'preqrequired', '__prerequired__']:
             kwargs.pop(ignore, None)
 
-        certs = __salt__['publish.publish'](
-            tgt=ca_server,
-            fun='x509.sign_remote_certificate',
-            arg=six.text_type(kwargs))
+        # If the provided argument is not a list, we assume it only contain one server
+        if not isinstance(ca_server, list):
+            ca_server = [ca_server]
+        random.shuffle(ca_server)
+
+        # This variable will contains some errors logs collected during the loop
+        # over the CA servers provided, it will be return if no certificate got generated
+        error_log = []
+        for server in ca_server:
+            certs = __salt__['publish.publish'](
+                tgt=server,
+                fun='x509.sign_remote_certificate',
+                arg=six.text_type(kwargs))
+            if certs is None or not any(certs):
+                error_log.append("{0}: ca_server did not respond "
+                                 "or salt master must permit peers to "
+                                 "call the sign_remote_certificate function.".format(server))
+            else:
+                # Sometimes, an error is returned instead of the certificate.
+
+                # If we provide multiple CA servers, we want to make it HA, so
+                # we have to detect that's not a certificate and try an other server.
+
+                # Also, we don't want to break in case of the cert_txt is a dict
+                # as it could be the output of the "dry-run" first request.
```

Comment:
Is there any other marker that it's the dry-run, or is it just that `certs[server]` isn't a string type?

Reply:
In case of the first "test" request, the certs look like this :
```
{'ca-2.lxd-dev-1.test.lxd': {'Version': 3, 'Key Size': 4096, 'Serial Number': '2E:CB:CF:AA:D7:D2:74:84', 'SHA-256 Finger Print': 'AE:AA:C7:A7:18:85:6E:CD:B0:80:F3:DA:B6:04:40:84:F0:77:95:73:AE:94:ED:87:B1:E8:DE:B4:2C:DC:13:9F', 'MD5 Finger Print': '38:80:43:27:21:9F:94:0B:DB:D0:07:1F:1A:87:29:B7', 'SHA1 Finger Print': '43:BC:06:9E:0C:FC:BA:5B:B9:C7:09:10:C9:77:AB:0C:37:C2:F5:AA', 'Subject': {'C': 'CA', 'SP': 'Quebec', 'L': 'Montreal', 'O': 'system:nodes', 'CN': 'system:node:k8s-node-1.lxd-dev-1'}, 'Subject Hash': '44:FC:7A:07', 'Issuer': {'C': 'CA', 'SP': 'Quebec', 'L': 'Montreal', 'CN': 'k8s.ca.myproject'}, 'Issuer Hash': 'E7:23:D9:24', 'Not Before': '2019-06-06 14:37:03', 'Not After': '2019-09-04 14:37:03', 'Public Key': '-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA2KaUCOUha6L44GahEyba\n8ymgcJnkjL7Ne+3jWODHOIHY9vKk4UUYn0Yj/sNmJ2UrjESriYrIS6OSsDEEzfyL\nbVZ2DfI+ECENdaltyIE8PHNgVkZnnGA4tSq+UOHtXNqogQhNLcR1W5VnvjpTcHSy\nd4XjzBmgMS/H4oYoFHtGvp6qh7Y++0Gj6cVesixG6yDwvLJvKSTieLmXjDJPoL2A\nNW40bb5qovfP6i1Nq40AcVmJWk4/RVF8bADP9dDkFDbx1Cv45Ol9a5/+eQ20do5U\neX46JWh0sgjeufI5W8he6OwNvEgnzy1aN56/WMv01EkRgjGQEQFvyNU8rNN5ubsk\nJ7XJMZgrs8Qzu2neDOmfc7oMcaSv1HjjYj/wE5F56/zRve7PDveQKUI65MXEVzlc\n+T1vmYBMJ3Zuo4gmKDptnfwTGjsLYB6uXWMfB85y3ZA2qyWGQ3Ml+ow5/iiodOfG\ny2myvNMSdYpb1ducQAbq+1pxiKBI304vu+IVthi3bsnBG+7L5oPj49IBr+5iRHUJ\nM4PqqOO+l9qUplYIoi93HVZ5hyulPYSAB+kaiXoOdUbieJ+nlaEjIpXYAnlO4YIp\ncplI/kdXx0IQADi4SgMHGL9/FZ2Q/XHTdxyKKHQH8Qz8efwL6Pf4wDJ2yAkk8rHi\ny4KKyiGCBMuBxuZzfQetOx0CAwEAAQ==\n-----END PUBLIC KEY-----\n', 'X509v3 Extensions': {'basicConstraints': 'critical CA:FALSE', 'keyUsage': 'critical Digital Signature, Key Encipherment, Key Agreement', 'subjectKeyIdentifier': 'BD:8C:15:9A:2F:4E:D5:BF:51:23:E5:74:3B:51:5B:40:B1:87:BD:45', 'authorityKeyIdentifier': 'keyid:D3:82:D2:44:D6:C5:FF:05:75:0D:90:61:E0:B4:CC:DF:3C:AB:9F:E9\nDirName:/C=CA/ST=Quebec/L=Montreal/CN=k8s.ca.myproject\nserial:7F:C1:3B:35:42:F3:E0:EC\n', 'subjectAltName': 'IP Address:10.0.6.197, DNS:k8s-node-1.lxd-dev-1.test.lxd'}, 'Issuer Public Key': '-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA1K+evRbRF1XECrDlAJUO\nilEu7RmFN57BVua98+Med+lQXki7q2GB6YVnfIcdt2WpPUTCoG0w2clqtTbcxc93\nmgkv1Qjc2XKwC2boj07QyFhgXlgAb7Anu2FaBjAGR15pAoTSHxreuU5Z923mMJWb\nrwSKA+/WijhP2wm9Pi+wQpQKiRD5a2MpDxLx3FAPpXXmPKk4pg8RNhuwOczibibL\nnsADcosMrVzxDnaJ0y19ENxa2LXuBZqSHt6fZyTjyILRdxnw4+0E+BWc61IODSMG\nzRnp6cSXwjr2ck/4cSZgeiVfBlQADt7rTDI8uGMBZGBRy4nCjjB+tEeJo/FRnPcJ\nHRbyScCoUqUpETDhfRaiTzxjqV9tqyG2euAFanvGejecxcrHzPV30i+PbOhuVBVX\nSmURkIXyXR2FKHYXEErucgi19ruAQtonFg0WC0lzDUlRax226NcIfJ/1xf0MqkkM\nbxXup1C5tt0gQBixkEq8716SibxiLohn0optiXPwbfq3Ufgbtj+Z606jP3+Ft8nq\nessAj20R+f+aYWVh0b0Wnttzwgqfwft1K2xbJRZYwl1qymXLB6mOgXylc+QLtNHE\nxOLwWEYoy9XuuY18LrYu0/E3U2EqwUqc4a+w0ooRo8Z8mKnEMwkzAai8cGujZvxf\nBb6AccTS/pokxEAEE17/kf8CAwEAAQ==\n-----END PUBLIC KEY-----\n'}}{'ca-2.lxd-dev-1.test.lxd': {'Version': 3, 'Key Size': 4096, 'Serial Number': '2E:CB:CF:AA:D7:D2:74:84', 'SHA-256 Finger Print': 'AE:AA:C7:A7:18:85:6E:CD:B0:80:F3:DA:B6:04:40:84:F0:77:95:73:AE:94:ED:87:B1:E8:DE:B4:2C:DC:13:9F', 'MD5 Finger Print': '38:80:43:27:21:9F:94:0B:DB:D0:07:1F:1A:87:29:B7', 'SHA1 Finger Print': '43:BC:06:9E:0C:FC:BA:5B:B9:C7:09:10:C9:77:AB:0C:37:C2:F5:AA', 'Subject': {'C': 'CA', 'SP': 'Quebec', 'L': 'Montreal', 'O': 'system:nodes', 'CN': 'system:node:k8s-node-1.lxd-dev-1'}, 'Subject Hash': '44:FC:7A:07', 'Issuer': {'C': 'CA', 'SP': 'Quebec', 'L': 'Montreal', 'CN': 'k8s.ca.myproject'}, 'Issuer Hash': 'E7:23:D9:24', 'Not Before': '2019-06-06 14:37:03', 'Not After': '2019-09-04 14:37:03', 'Public Key': '-----BEGIN PUBLIC KEY-----\nXXXXXXXXX\n-----END PUBLIC KEY-----\n', 'X509v3 Extensions': {'basicConstraints': 'critical CA:FALSE', 'keyUsage': 'critical Digital Signature, Key Encipherment, Key Agreement', 'subjectKeyIdentifier': 'BD:8C:15:9A:2F:4E:D5:BF:51:23:E5:74:3B:51:5B:40:B1:87:BD:45', 'authorityKeyIdentifier': 'keyid:D3:82:D2:44:D6:C5:FF:05:75:0D:90:61:E0:B4:CC:DF:3C:AB:9F:E9\nDirName:/C=CA/ST=Quebec/L=Montreal/CN=k8s.ca.myproject\nserial:7F:C1:3B:35:42:F3:E0:EC\n', 'subjectAltName': 'IP Address:10.0.6.197, DNS:k8s-node-1.lxd-dev-1.test.lxd'}, 'Issuer Public Key': '-----BEGIN PUBLIC KEY-----\nXXXXXX\n-----END PUBLIC KEY-----\n'}}
```

In case of it need to generate/regenerate a certificate, certs will look like this:

```
{'ca-2.lxd-dev-1.test.lxd': '-----BEGIN CERTIFICATE-----\nXXXXXXXX\n-----END CERTIFICATE-----\n'}
```

In of error, we can get something like this: 
```
{'ca-1.lxd-dev-1.test.lxd': "[Errno 1] Operation not permitted: '/etc/pki/k8s/issued_certs/7D:59:EF:E1:17:01:34:78.crt'"}
```
In case of the error, the states is going to return an error exception raised by "get_pem_entry" complaining that the certificate is invalid but it will write it in the file even if it's not a valid file.